In [1]:
import pandas as pd
import numpy as np

In [2]:
file_path = r'C:\Users\19802\Documents\学习\Cornell_AB Testing_25Fall\jcb540s_tool-coupon.xlsx'
sheet_name = 'coupon'
coupon = pd.read_excel(r'C:\Users\19802\Documents\学习\Cornell_AB Testing_25Fall\jcb540s_tool-coupon.xlsx', sheet_name)

In [3]:
print(coupon.head())

   customerid group  visit  pageviews  duration  redeem  purchase  amount  \
0   100003171  CG_H      0          0         0       0         0     0.0   
1   100005146  T2_H      0          0         0       0         0     0.0   
2   100010245  T2_H      0          0         0       0         0     0.0   
3   100016186  T1_H      0          0         0       0         0     0.0   
4   100016657  T1_L      0          0         0       0         0     0.0   

   discount  last  numpurch  totdol  gender  age  
0       0.0   138         3  110.60       0   54  
1       0.0   328         1   94.60       1   51  
2       0.0   349         3  175.00       1   47  
3       0.0   122         5  314.23       1   45  
4       0.0     9         4   48.91       1   44  


In [4]:
#Create a new column 'coupon' to indicate better coupon= 2, base coupon =1, no coupon =0
coupon['coupon'] = coupon['group'].apply(
    lambda x: 1 if x in ['T1_L', 'T1_H'] 
    else 2 if x in ['T2_H'] 
    else 0
)

print(coupon.head())

   customerid group  visit  pageviews  duration  redeem  purchase  amount  \
0   100003171  CG_H      0          0         0       0         0     0.0   
1   100005146  T2_H      0          0         0       0         0     0.0   
2   100010245  T2_H      0          0         0       0         0     0.0   
3   100016186  T1_H      0          0         0       0         0     0.0   
4   100016657  T1_L      0          0         0       0         0     0.0   

   discount  last  numpurch  totdol  gender  age  coupon  
0       0.0   138         3  110.60       0   54       0  
1       0.0   328         1   94.60       1   51       2  
2       0.0   349         3  175.00       1   47       2  
3       0.0   122         5  314.23       1   45       1  
4       0.0     9         4   48.91       1   44       1  


In [5]:
#Transform to dummy variable
#Create a new column 'high' to indicate high-value segment
coupon['high'] = coupon['group'].apply(
    lambda x: 1 if x in ['T1_H', 'T2_H', 'CG_H']
    else 0
)

print(coupon.head())

   customerid group  visit  pageviews  duration  redeem  purchase  amount  \
0   100003171  CG_H      0          0         0       0         0     0.0   
1   100005146  T2_H      0          0         0       0         0     0.0   
2   100010245  T2_H      0          0         0       0         0     0.0   
3   100016186  T1_H      0          0         0       0         0     0.0   
4   100016657  T1_L      0          0         0       0         0     0.0   

   discount  last  numpurch  totdol  gender  age  coupon  high  
0       0.0   138         3  110.60       0   54       0     1  
1       0.0   328         1   94.60       1   51       2     1  
2       0.0   349         3  175.00       1   47       2     1  
3       0.0   122         5  314.23       1   45       1     1  
4       0.0     9         4   48.91       1   44       1     0  


In [6]:
#Transform to dummy variable
#Create new columns 'base' and 'better' as dummy variables
coupon['base'] = coupon['group'].apply(
    lambda x: 1 if x in ['T1_L', 'T1_H']
    else 0
)

coupon['better'] = coupon['group'].apply(
    lambda x: 1 if x in ['T2_H']
    else 0
)

print(coupon.head())

   customerid group  visit  pageviews  duration  redeem  purchase  amount  \
0   100003171  CG_H      0          0         0       0         0     0.0   
1   100005146  T2_H      0          0         0       0         0     0.0   
2   100010245  T2_H      0          0         0       0         0     0.0   
3   100016186  T1_H      0          0         0       0         0     0.0   
4   100016657  T1_L      0          0         0       0         0     0.0   

   discount  last  numpurch  totdol  gender  age  coupon  high  base  better  
0       0.0   138         3  110.60       0   54       0     1     0       0  
1       0.0   328         1   94.60       1   51       2     1     0       1  
2       0.0   349         3  175.00       1   47       2     1     0       1  
3       0.0   122         5  314.23       1   45       1     1     1       0  
4       0.0     9         4   48.91       1   44       1     0     1       0  


In [7]:
#Create a low_value seg as df
coupon['high'].dtype

low_value_segment = coupon[coupon['high'] == 0]

print(low_value_segment)

       customerid group  visit  pageviews  duration  redeem  purchase  amount  \
4       100016657  T1_L      0          0         0       0         0     0.0   
5       100021299  T1_L      1          2       175       0         0     0.0   
8       100024643  T1_L      0          0         0       0         0     0.0   
9       100028202  CG_L      0          0         0       0         0     0.0   
11      100028818  T1_L      0          0         0       0         0     0.0   
...           ...   ...    ...        ...       ...     ...       ...     ...   
12949   125469864  CG_L      1          4        70       0         0     0.0   
12951   125474475  CG_L      0          0         0       0         0     0.0   
12955   125538236  T1_L      0          0         0       0         0     0.0   
12956   125563783  CG_L      0          0         0       0         0     0.0   
12958   125585750  T1_L      0          0         0       0         0     0.0   

       discount  last  nump

In [8]:
coupon['totdol'].describe()

count    12959.000000
mean        97.677053
std         75.443554
min         17.600000
25%         33.300000
50%         77.900000
75%        136.825000
max        381.320000
Name: totdol, dtype: float64

In [9]:
#Get summary statistics for the column 'totdol' group by base coupon or no coupon for low_value segment
low_value_segment.groupby('base')['totdol'].describe()

count       mean        std   min     25%   50%     75%   max
base                                                                
0     1998.0  39.877735  17.660587  18.6  25.600  35.0  51.150  87.2
1     4787.0  40.301959  18.166353  17.6  25.375  34.6  51.875  87.4

In [ ]:
high_value_segment = coupon[coupon['high'] == 1]

**Verify Random Assignment for test vs control group for all the pre-test variables**

*Low_Value Segment*

In [10]:
#Verify Random Assignment for test vs control group for all the pre-test variables 
from scipy import stats

#Low_value segment
#1.totdol - total spend in past 12 months
#Null hypothesis: there is no relationship between group assignment (test/control) and spending patterns.
control_totdol = low_value_segment[low_value_segment['base'] == 0]['totdol']
test_totdol = low_value_segment[low_value_segment['base'] == 1]['totdol']

t_stat, p_value = stats.ttest_ind(control_totdol, test_totdol, equal_var=True)  # set equal_var=False for Welch’s t-test
print("t-statistic:", t_stat)
print("p-value:", p_value)

t-statistic: -0.8839368755236597
p-value: 0.3767616186103123


P value bigger than 0.5 means that the null hypothesis cannot be rejected, which means randomization was achieved in the low-value segment between test and control groups in terms of totdor.

In [11]:
#2. last - number of days since last purchase
control_last = low_value_segment[low_value_segment['base'] == 0]['last']
test_last = low_value_segment[low_value_segment['base'] == 1]['last']

t_stat, p_value = stats.ttest_ind(control_last, test_last, equal_var = True)
print('t-statistics:', t_stat)
print('p-value:', p_value)

t-statistics: 1.4093428098328973
p-value: 0.15877966942962546


P value is 0.159, which is bigger than 0.5, we cannot reject null hypothesis, which means randomization was achieved in the low-value segment between test and control groups in terms of number of days since last purchase.

In [12]:
#3. numpurch - number of purchases in past 12 months
control_numpurch = low_value_segment[low_value_segment['base'] == 0]['numpurch']
test_numpurch = low_value_segment[low_value_segment['base'] == 1]['numpurch']

t_stat, p_value = stats.ttest_ind(control_numpurch, test_numpurch, equal_var = True)
print('t-statistics:', t_stat)
print('p-value:', p_value)

t-statistics: -1.2957735436088202
p-value: 0.19509760653807492


P value is 0.195, which is bigger than 0.05. We cannot reject the null hypothesis, which means randomization was achieved in the low-value segment between test and control groups in terms of number of purchases in the past 12 months.

In [13]:
#4. gender
control_gender = low_value_segment[low_value_segment['base'] == 0]['gender']
test_gender = low_value_segment[low_value_segment['base'] == 1]['gender']

t_stat, p_value = stats.ttest_ind(control_gender, test_gender, equal_var = True)
print('t-statistic:', t_stat)
print('p-value:', p_value)

t-statistic: 0.602356631378536
p-value: 0.5469569068782603


P value is 0.54, which is higher than 0.05. We cannot reject null hypothesis, which means randomization was achieved in the low-value segment between test and control groups in terms of gender.

In [14]:
#5. age
control_age = low_value_segment[low_value_segment['base'] == 0]['age']
test_age = low_value_segment[low_value_segment['base'] == 1]['age']

t_stat, p_value = stats.ttest_ind(control_age, test_age, equal_var = True)
print('t-statistic:', t_stat)
print('p-value:', p_value)

t-statistic: -2.0540811004385566
p-value: 0.04000608354890501


? P value is 0.04, which is lower than 0.05. We are rejecting null hypothesis, which means randomization was NOT achieved in the low-value segment between test and control groups in terms of age.

*High_Value Segment*

In [15]:
high_value_segment = coupon[coupon['high'] == 1]

print(high_value_segment)

       customerid group  visit  pageviews  duration  redeem  purchase  amount  \
0       100003171  CG_H      0          0         0       0         0     0.0   
1       100005146  T2_H      0          0         0       0         0     0.0   
2       100010245  T2_H      0          0         0       0         0     0.0   
3       100016186  T1_H      0          0         0       0         0     0.0   
6       100022013  T2_H      0          0         0       0         0     0.0   
...           ...   ...    ...        ...       ...     ...       ...     ...   
12950   125473785  T2_H      1          2        96       0         0     0.0   
12952   125486705  T1_H      0          0         0       0         0     0.0   
12953   125507576  T2_H      0          0         0       0         0     0.0   
12954   125526986  T1_H      0          0         0       0         0     0.0   
12957   125568879  T1_H      1          3        74       0         0     0.0   

       discount  last  nump

In [16]:
print(high_value_segment)

       customerid group  visit  pageviews  duration  redeem  purchase  amount  \
0       100003171  CG_H      0          0         0       0         0     0.0   
1       100005146  T2_H      0          0         0       0         0     0.0   
2       100010245  T2_H      0          0         0       0         0     0.0   
3       100016186  T1_H      0          0         0       0         0     0.0   
6       100022013  T2_H      0          0         0       0         0     0.0   
...           ...   ...    ...        ...       ...     ...       ...     ...   
12950   125473785  T2_H      1          2        96       0         0     0.0   
12952   125486705  T1_H      0          0         0       0         0     0.0   
12953   125507576  T2_H      0          0         0       0         0     0.0   
12954   125526986  T1_H      0          0         0       0         0     0.0   
12957   125568879  T1_H      1          3        74       0         0     0.0   

       discount  last  nump

In [17]:
#High_value segment
#1.totdol - total spend in past 12 months
#Null hypothesis: there is no relationship between group assignment (test/control) and spending patterns.
#Since there are two test groups (base & better) and one control group, need to have 3 T tests.
control_h_totdol = high_value_segment[(high_value_segment['base'] == 0) & (high_value_segment['coupon'] != 2)]['totdol']
test_h_totdol = high_value_segment[(high_value_segment['base'] == 1) & (high_value_segment['coupon'] != 2)]['totdol']

t_stat, p_value = stats.ttest_ind(control_h_totdol, test_h_totdol, equal_var=True)  # set equal_var=False for Welch’s t-test
print("t-statistic:", t_stat)
print("p-value:", p_value)

t-statistic: 0.666641320740972
p-value: 0.5050576850497586


**Linear Regression**

In [18]:
import statsmodels.api as sm

In [19]:
#Low_value segment
#Is there a relationsihp between receiving the base coupon and purchase amount?
#base vs amount

X = low_value_segment[['base']]
y = low_value_segment['amount']

# Add constant for intercept
X = sm.add_constant(X)

# Fit OLS model
linear_low= sm.OLS(y, X).fit()

# Print summary
print(linear_low.summary())

                            OLS Regression Results                            
Dep. Variable:                 amount   R-squared:                       0.002
Model:                            OLS   Adj. R-squared:                  0.001
Method:                 Least Squares   F-statistic:                     11.17
Date:                Mon, 27 Oct 2025   Prob (F-statistic):           0.000835
Time:                        06:47:47   Log-Likelihood:                -20752.
No. Observations:                6785   AIC:                         4.151e+04
Df Residuals:                    6783   BIC:                         4.152e+04
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.2546      0.115      2.208      0.0

1. Test the overall significance of the regression model compared to a model with no independent varilabes.
   The P-value for the model is 0.000835 --> Reject null, meaning that the regression model could be valuable.
2. Assess how good the regression model is.
   Adjusted R^2 is 0.001 --> The regression model can explain about 0.1% of the information in the dependent variable.
3. Test the significance of each individual coefficient
   the coef is 0.4588, and p value is 0.001 --> reject null, meaning the base coupon is a significant predictor of purchase amount resulting in a lift of $0.46.
4. Make predictions.
   

In [20]:
# Make predictions (like R's predict(..., type='response'))
amount_hat = linear_low.predict(X)

# Summary statistics (like R's summary(amount_hat))
print(pd.Series(amount_hat).describe())

count    6785.000000
mean        0.578276
std         0.209136
min         0.254585
25%         0.254585
50%         0.713378
75%         0.713378
max         0.713378
dtype: float64


In [21]:
#High_value segment
#Is there a relationsihp between receiving the base coupon and purchase amount?
#base vs amount

X = high_value_segment[['base', 'better']]
y = high_value_segment['amount']

# Add constant for intercept
X = sm.add_constant(X)

# Fit OLS model
linear_high= sm.OLS(y, X).fit()

# Print summary
print(linear_high.summary())

                            OLS Regression Results                            
Dep. Variable:                 amount   R-squared:                       0.002
Model:                            OLS   Adj. R-squared:                  0.002
Method:                 Least Squares   F-statistic:                     6.990
Date:                Mon, 27 Oct 2025   Prob (F-statistic):           0.000928
Time:                        06:51:32   Log-Likelihood:                -24901.
No. Observations:                6174   AIC:                         4.981e+04
Df Residuals:                    6171   BIC:                         4.983e+04
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          1.1597      0.483      2.403      0.0

1. Test the overall significance of the regression model compared to a model with no independent varilabes.
   The P-value for the model is 0.000928 --> Reject null, meaning that the regression model is significant.
2. Assess how good the regression model is.
   Adjusted R^2 is 0.002 --> The regression model can explain about 0.2% of the information in the dependent variable.
3. Test the significance of each individual coefficient
   the coef for base coupon is 1.36 and for better coupon is 1.97, and p values are 0.016 and 0.018 -> reject null, meaning both coupon choices are effective in increasing purchase amount(a lift of 0.36 and 1.97) in the high-value sgment and the better coupon is more effective than the base coupon.
4. Make predictions.

In [23]:
# Make predictions (like R's predict(..., type='response'))
amount_hat_high = linear_high.predict(X)

# Summary statistics (like R's summary(amount_hat))
print(pd.Series(amount_hat).describe())

count    6174.000000
mean        2.686494
std         0.650076
min         1.159719
25%         2.521471
50%         3.133664
75%         3.133664
max         3.133664
dtype: float64


**Question: for low value segment, is there still a relationship between receiving the base coupon and purchase amount once you account for other customer characteristics?**


In [27]:
X = low_value_segment[['base', 'last', 'numpurch','totdol', 'gender', 'age']]
y = low_value_segment['amount']

X = sm.add_constant(X)

linear2_low = sm.OLS(y, X).fit()

print(linear2_low.summary())

                            OLS Regression Results                            
Dep. Variable:                 amount   R-squared:                       0.005
Model:                            OLS   Adj. R-squared:                  0.004
Method:                 Least Squares   F-statistic:                     5.154
Date:                Mon, 27 Oct 2025   Prob (F-statistic):           2.69e-05
Time:                        07:02:30   Log-Likelihood:                -20742.
No. Observations:                6785   AIC:                         4.150e+04
Df Residuals:                    6778   BIC:                         4.155e+04
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.4304      0.403     -1.069      0.2

1. Test the overall significance of the regression model compared to a model with no independent varilabes. The P-value for the model is 2.69e-05 --> Reject null, meaning that the regression model is significant.
2. Assess how good the regression model is. Adjusted R^2 is 0.004 --> The regression model can explain about 0.4% of the information in the dependent variable.
3. Test the significance of each individual coefficient --> base and last are the only two independent variables who have less than 0.05 p value, meaning the effect of the base coupon and number of days since last purchase for low value segment are significant. I am ignoring age since age is not randomized based my previous analysis.
4. Make predictions

In [29]:
# Make predictions (like R's predict(..., type='response'))
amount_hat2_low = linear2_low.predict(X)

# Summary statistics (like R's summary(amount_hat))
print(pd.Series(amount_hat2_low).describe())

count    6785.000000
mean        0.578276
std         0.347591
min        -0.485545
25%         0.348793
50%         0.581417
75%         0.813294
max         1.967852
dtype: float64


**Develop a linear regression model for the high-value segment, using base, coupon, last, numpurch, totdol. Which variables are significant predictor of purchase amount?**

In [30]:
x = high_value_segment[['base','better','last','numpurch','totdol']]
y = high_value_segment['amount']

x = sm.add_constant(x)

linear2_high = sm.OLS(y, x).fit()

print(linear2_high.summary())

                            OLS Regression Results                            
Dep. Variable:                 amount   R-squared:                       0.016
Model:                            OLS   Adj. R-squared:                  0.015
Method:                 Least Squares   F-statistic:                     19.49
Date:                Mon, 27 Oct 2025   Prob (F-statistic):           2.60e-19
Time:                        07:15:09   Log-Likelihood:                -24860.
No. Observations:                6174   AIC:                         4.973e+04
Df Residuals:                    6168   BIC:                         4.977e+04
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.4465      0.760     -0.588      0.5

In [33]:
X = high_value_segment[['base', 'better', 'last', 'numpurch','totdol','gender', 'age']]
y = high_value_segment['amount']

X = sm.add_constant(X)

linear3_high = sm.OLS(y, X).fit()

print(linear3_high.summary())

                            OLS Regression Results                            
Dep. Variable:                 amount   R-squared:                       0.016
Model:                            OLS   Adj. R-squared:                  0.015
Method:                 Least Squares   F-statistic:                     14.67
Date:                Mon, 27 Oct 2025   Prob (F-statistic):           4.40e-19
Time:                        07:16:34   Log-Likelihood:                -24857.
No. Observations:                6174   AIC:                         4.973e+04
Df Residuals:                    6166   BIC:                         4.978e+04
Df Model:                           7                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.4387      1.075     -0.408      0.6

In [34]:
x = high_value_segment[['base','better']]
y = high_value_segment['amount']

x = sm.add_constant(x)

linear4_high = sm.OLS(y, x).fit()

print(linear4_high.summary())

                            OLS Regression Results                            
Dep. Variable:                 amount   R-squared:                       0.002
Model:                            OLS   Adj. R-squared:                  0.002
Method:                 Least Squares   F-statistic:                     6.990
Date:                Mon, 27 Oct 2025   Prob (F-statistic):           0.000928
Time:                        07:42:11   Log-Likelihood:                -24901.
No. Observations:                6174   AIC:                         4.981e+04
Df Residuals:                    6171   BIC:                         4.983e+04
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          1.1597      0.483      2.403      0.0